src/pudl_archiver/archiver/classes.py

```diff
@@ -0,0 +1,165 @@
+        partitions. What this means in practice is calling an `async` function and yielding
+        the results without awaiting them. This allows the base class to gather all of these
+        awaitables and download the resources concurrently.
+        """
+        ...
+
+    async def download_zipfile(
+        self, url: str, file: Path | io.BytesIO, retries: int = 5, **kwargs
+    ):
+        """Attempt to download a zipfile and retry if zipfile is invalid.
+
+        Args:
+            url: URL of zipfile.
+            file: Local path to write file to disk or bytes object to save file in memory.
+            retries: Number of times to attempt to download a zipfile.
+            kwargs: Key word args to pass to request.
+        """
+        for _ in range(0, retries):
+            await self.download_file(url, file, **kwargs)
+
+            if zipfile.is_zipfile(file):
+                return None
+
+        # If it makes it here that means it couldn't download a valid zipfile
+        raise RuntimeError(f"Failed to download valid zipfile from {url}")
+
+    async def download_file(self, url: str, file: Path | io.BytesIO, **kwargs):
+        """Download a file using async session manager.
+
+        Args:
+            url: URL to file to download.
+            file: Local path to write file to disk or bytes object to save file in memory.
+            kwargs: Key word args to pass to request.
+        """
+        async with self.session.get(url, **kwargs) as response:
+            if isinstance(file, Path):
+                with open(file, "wb") as f:
+                    f.write(await response.read())
+            elif isinstance(file, io.BytesIO):
+                file.write(await response.read())
+
+    async def get_hyperlinks(
+        self,
+        url: str,
+        filter_pattern: typing.Pattern | None = None,
+        verify: bool = True,
+    ) -> list[str]:
+        """Return all hyperlinks from a specific web page.
+
+        This is a helper function to perform very basic web-scraping functionality.
+        It extracts all hyperlinks from a web page, and returns those that match
+        a specified pattern. This means it can find all hyperlinks that look like
+        a download link to a single data resource.
+
+        Args:
+            url: URL of web page.
+            filter_pattern: If present, only return links that contain pattern.
+            verify: Verify ssl certificate (EPACEMS https source has bad certificate).
+        """
+        # Parse web page to get all hyperlinks
+        parser = _HyperlinkExtractor()
+        async with self.session.get(url, ssl=verify) as response:
+            text = await response.text()
+            parser.feed(text)
+
+        # Filter to those that match filter_pattern
+        hyperlinks = parser.hyperlinks
+        if filter_pattern:
+            hyperlinks = {link for link in hyperlinks if filter_pattern.search(link)}
+
+        return hyperlinks
```

Comment:
Does it make sense to raise an exception or warning if the regex search doesn't find any matches? Something like "The archiver couldn't find any hyperlinks that match {pattern}. Make sure your `filter_pattern` is correct or if the structure of the {url} page changed."

Reply:
Good call, added the message. This should give a warning for the comment above too.